conda_kapsel/project_commands.py

```diff
@@ -28,6 +29,114 @@ def _is_windows():
     return (platform.system() == 'Windows')
 
 
+_ArgSpec = namedtuple('_ArgSpec', ['option', 'has_value'])
+
+_http_specs = (_ArgSpec('--kapsel-host', True), _ArgSpec('--kapsel-port', True), _ArgSpec('--kapsel-url-prefix', True),
+               _ArgSpec('--kapsel-no-browser', False))
+
+
+class _ArgsTransformer(object):
+    def __init__(self, specs):
+        self.specs = specs
+
+    def _parse_args_removing_known(self, results, args):
+        if not args:
+            return []
+
+        arg = args[0]
+
+        if arg == '--':
+            return args
+
+        for spec in self.specs:
+            if spec.has_value:
+                with_equals = spec.option + "="
+                if arg == spec.option:
+                    if len(args) == 1 or args[1].startswith("-"):
+                        # This means there isn't a value for the option, which
+                        # ideally might report a syntax error, but we have to
+                        # do a significant refactoring to get a syntax error
+                        # up to a user-visible place from here. Hopefully
+                        # the command we are launching will complain about the
+                        # empty value.
+                        results[spec.option].append('')
+                        return self._parse_args_removing_known(results, args[1:])
+                    else:
+                        results[spec.option].append(args[1])
+                        return self._parse_args_removing_known(results, args[2:])
+                elif arg.startswith(with_equals):
+                    results[spec.option].append(arg[len(with_equals):])
+                    return self._parse_args_removing_known(results, args[1:])
+            elif arg == spec.option:
+                results[spec.option] = [True]
+                return self._parse_args_removing_known(results, args[1:])
+
+        return [arg] + self._parse_args_removing_known(results, args[1:])
+
+    def transform_args(self, args):
+        results = {spec.option: [] for spec in self.specs}
+        with_removed = self._parse_args_removing_known(results, args)
+        # flatten results with deterministic sort to ease testing
+        results_list = sorted(results.items(), key=lambda x: x[0])
+        return self.add_args(results_list, with_removed)
+
+    def add_args(self, results, args):
+        raise RuntimeError("not implemented")  # pragma: no cover
+
+
+class _BokehArgsTransformer(_ArgsTransformer):
+    def __init__(self):
+        super(_BokehArgsTransformer, self).__init__(_http_specs)
+
+    def add_args(self, results, args):
+        added = []
+        for (option, values) in results:
+            if option in ('--kapsel-host', '--kapsel-port'):
+                for v in values:
+                    added.append(option.replace('kapsel-', ''))
+                    added.append(v)
```

Comment:
I supported multiple copies of every option (such as --kapsel-host), but it only makes sense for some of them (such as host) and not for others (such as port). I figure let the invoked tool such as bokeh complain about that. Also I should add test coverage for the multiple instances case.